addons/mrp/models/stock_rule.py

```diff
@@ -155,16 +158,16 @@ def _prepare_mo_vals(self, product_id, product_qty, product_uom, location_dest_i
             'never_product_template_attribute_value_ids': values.get('never_product_template_attribute_value_ids'),
             'product_qty': product_uom._compute_quantity(product_qty, bom.product_uom_id) if bom else product_qty,
             'product_uom_id': bom.product_uom_id.id if bom else product_uom.id,
-            'location_src_id': self.picking_type_id.default_location_src_id.id,
-            'location_dest_id': self.picking_type_id.default_location_dest_id.id or location_dest_id.id,
+            'location_src_id': bom.picking_type_id.default_location_src_id.id or self.picking_type_id.default_location_src_id.id,
```

Comment:
To make things easier to read, maybe we could have a:
```py
picking_type = bom.picking_type_id or self.picking_type_id
```
and then use that `picking_type`?